src/python/WMCore/Services/DBS/DBSUtils.py

```diff
@@ -0,0 +1,149 @@
+    :param blocks: list of blocks
+    :param logger: logger
+    :return: list of file lumis
+    """
+    urls = ['%s/filelumis?block_name=%s' % (dbsUrl, b.replace('#', '%23')) for b in blocks]
+    func = aggFileLumis
+    uKey = 'block_name'
+    return getUrls(urls, func, uKey, logger)
+
+
+def dbsBlockOrigin(dbsUrl, blocks, logger=None):
+    """
+    Concurrent counter part of DBS files API
+
+    :param dbsUrl: DBS URL
+    :param blocks: list of blocks
+    :param logger: logger
+    :return: list of block origins for a given parent lfns
+    """
+    urls = ['%s/blockorigin?block_name=%s' % (dbsUrl, b.replace('#', '%23')) for b in blocks]
+    func = None
+    uKey = 'block_name'
+    return getUrls(urls, func, uKey, logger)
+
+
+def dbsParentFilesGivenParentDataset(dbsUrl, parentDataset, fInfo, logger=None):
+    """
+    Obtain parent files for given fileInfo object
+
+    :param dbsUrl: DBS URL
+    :param parentDataset: parent dataset name
+    :param fInfo: file info object
+    :param logger: logger object
+    :return: list of parent files for given file info object
+    """
+    urls = []
+    for fileInfo in fInfo:
+        run = fileInfo['run_num']
+        lumis = urllib.parse.quote_plus(str(fileInfo['lumi_section_num']))
+        url = '%s/files?dataset=%s&run_num=%s&lumi_list=%s' \
+                % (dbsUrl, parentDataset, run, lumis)
+        if ':8443' not in url:
+            url = url.replace("cern.ch", "cern.ch:8443")
+        urls.append(url)
+    func = None
+    uKey = None
+    rdict = getUrls(urls, func, uKey, logger)
+    parentFiles = defaultdict(set)
+    for fileInfo in fInfo:
+        run = fileInfo['run_num']
+        lumis = urllib.parse.quote_plus(str(fileInfo['lumi_section_num']))
+        url = '%s/files?dataset=%s&run_num=%s&lumi_list=%s' \
+                % (dbsUrl, parentDataset, run, lumis)
+        if ':8443' not in url:
+            url = url.replace("cern.ch", "cern.ch:8443")
+        if url in rdict:
+            pFileList = rdict[url]
+            pFiles = set([x['logical_file_name'] for x in pFileList])
+            parentFiles[fileInfo['logical_file_name']] = \
+                    parentFiles[fileInfo['logical_file_name']].union(pFiles)
+    return parentFiles
+
+
+def getUrls(urls, aggFunc, uKey=None, logger=None):
+    """
+    Perform parallel DBS calls for given set of urls and apply given aggregation
+    function to the results.
+
+    :param urls: list of DBS urls to call
+    :param aggFunc: aggregation function
+    :param uKey: url parameter to use for final dictionary
+    :param logger: logger object
+    :return: dictionary of resuls where keys are urls and values are obtained results
+    """
+    data = multi_getdata(urls, ckey(), cert())
+
+    rdict = {}
+    for row in data:
+        url = row['url']
+        if uKey:
+            key = urlParams(url).get(uKey)
+        else:
+            key = url
+        code = int(row.get('code', 200))
+        if code != 200:
```

Comment:
Shouldn't we consider fail http requests only if the status code is >= 400?

Reply:
this is an edge use-case since non 200 codes does not return any data and may return plain string, e.g. the code 301 will return "Moved permanently", and code below explicitly extracts the `data` which is constructed by `multi_getdata` function from `pycurl_manager`. I would leave it as is because it explicitly will show non-data message.

Reply:
If my memory doesn't fail me, I've seen cases where we get successfully get actual data and the return code is != 200. Perhaps we should then consider at least codes 200-200 as successful? From:
https://developer.mozilla.org/en-US/docs/Web/HTTP/Status